tracee-ebpf/tracee/tracee.bpf.c

```diff
@@ -732,6 +729,21 @@ static __always_inline const char * get_cgroup_dirname(struct cgroup *cgrp)
     return READ_KERN(kn->name);
 }
 
+static __always_inline const u64 get_cgroup_id(struct cgroup *cgrp)
+{
+    struct kernfs_node *kn = READ_KERN(cgrp->kn);
+
+    if (kn == NULL)
+        return 0;
+
+    // id was not always of type u64 (but "union kernfs_node_id" before 5.5)
+    // yet we can read it as u64 as the size of both types is the same
+    u64 id;
+    bpf_probe_read(&id, sizeof(u64), GET_FIELD_ADDR(kn->id));
```

Comment:
I think you will have to add type flavor for v5.4.y here so the field relocation is correctly calculated on these kernels.

libbpf was unable to relocate this when loading the object in an v5.4 kernel, it poisoned the instruction so the eBPF verifier would make it fail:

```
; if (kn == NULL)
318: (15) if r3 == 0x0 goto pc+7
 R0=inv(id=0) R3_w=inv(id=0) R6=invP0 R7=inv(id=0) R8=inv(id=0) R9=inv2344 R10=fp0 fp-8=???????m fp-16=mmmmmmmm fp-24=mmmmmmmm fp-32=mmmmmmmm fp-40=mmmmmmmm fp-48=mmmmmmmm fp-56=mmmmmmmm fp-64=mmmmmmmm fp-72=mmmmmmmm fp-80=0000mmmm fp-88=map_value fp-96=ctx fp-104=00000000 fp-112=00000000 fp-120=0000mmmm fp-128=mmmmmmmm fp-136=mmmmmmmm fp-144=mmmmmmmm fp-152=mmmmmmmm fp-160=mmmmmmmm fp-168=mmmmmmmm fp-176=mmmmmmmm fp-184=mmmmmmmm fp-192=mmmmmmmm fp-200=mmmmmmmm fp-208=mmmmmmmm fp-216=ctx fp-224=mmmmmmmm fp-232=mmmmmmmm
319: (85) call unknown#195896080
invalid func unknown#195896080
processed 319 insns (limit 1000000) max_states_per_insn 0 total_states 18 peak_states 18 mark_read 16
```

Reply:
The 5.4 kernel has:

```
[1415] STRUCT 'kernfs_node' size=128 vlen=13
        'count' type_id=79 bits_offset=0
        'active' type_id=79 bits_offset=32
        'parent' type_id=1416 bits_offset=64
        'name' type_id=5 bits_offset=128
        'rb' type_id=559 bits_offset=192
        'ns' type_id=253 bits_offset=384
        'hash' type_id=9 bits_offset=448
        '(anon)' type_id=1424 bits_offset=512
        'priv' type_id=103 bits_offset=768
        'id' type_id=1423 bits_offset=832
        'flags' type_id=19 bits_offset=896
        'mode' type_id=59 bits_offset=912
        'iattr' type_id=1425 bits_offset=960

[1423] UNION 'kernfs_node_id' size=8 vlen=2
        '(anon)' type_id=1422 bits_offset=0
        'id' type_id=36 bits_offset=0
```

Latest kernels have:

```
[1418] STRUCT 'kernfs_node' size=128 vlen=13
        'count' type_id=71 bits_offset=0
        'active' type_id=71 bits_offset=32
        'parent' type_id=1419 bits_offset=64
        'name' type_id=4 bits_offset=128
        'rb' type_id=354 bits_offset=192
        'ns' type_id=86 bits_offset=384
        'hash' type_id=7 bits_offset=448
        '(anon)' type_id=1425 bits_offset=512
        'priv' type_id=85 bits_offset=768
        'id' type_id=30 bits_offset=832
        'flags' type_id=16 bits_offset=896
        'mode' type_id=53 bits_offset=912
        'iattr' type_id=1426 bits_offset=960

[30] TYPEDEF 'u64' type_id=22
```

I think what happened is that libbpf was unable to do this relocation because it had no other "flavor" of "struct kernfs_node" (containing the union kernfs_node_id) to work with (and generate the correct RELO code within .BTF.ext ELF section). WDYT ?

Reply:
BTW, for this new struct flavor you will likely need: https://github.com/aquasecurity/tracee/issues/1133 work to be done. I'm working on that now.

Reply:
@rafaeldtinoco can you have a look at https://github.com/aquasecurity/tracee/pull/1130/commits/10da2387c8065154baa5b18327d732feaaf6fcb9?
Hopefully, this will solve the problem for older kernels with BTF.

The problem is that I'm not sure btfgen knows how to handle type checks like I had to do here...

Reply:
Still working on the verifier thing, but Id like to clarify this finding... 

kernel 5.13 ID value for a single container being started and stopped:
```
$ sudo bpftool prog tracelog
         systemd-1       [000] d... 630056.825108: bpf_trace_printk: BIGGER ID = 31687
         systemd-1       [007] d... 630060.909689: bpf_trace_printk: BIGGER ID = 31687
```

kernels 5.4 ID value for a single container being started:
```
$ sudo cat /sys/kernel/debug/tracing/trace_pipe
            runc-9391    [000] d... 292082.860162: 0: SMALLER ID = 4294967462
            runc-9391    [000] d... 292082.861525: 0: SMALLER ID = 4294967883
            runc-9391    [000] d... 292082.861934: 0: SMALLER ID = 4294969343
            runc-9391    [000] d... 292082.862313: 0: SMALLER ID = 4294968684
            runc-9391    [000] d... 292082.862822: 0: SMALLER ID = 4294968008
            runc-9391    [000] d... 292082.863143: 0: SMALLER ID = 4294968321
            runc-9391    [000] d... 292082.863479: 0: SMALLER ID = 4294967416
            runc-9391    [000] d... 292082.863840: 0: SMALLER ID = 4294967371
            runc-9391    [000] d... 292082.864355: 0: SMALLER ID = 4294967344
            runc-9391    [000] d... 292082.865515: 0: SMALLER ID = 4294967368
            runc-9391    [000] d... 292082.865784: 0: SMALLER ID = 4294967774
            runc-9466    [003] d... 292090.195132: 0: SMALLER ID = 4294968684
            runc-9466    [003] d... 292090.195269: 0: SMALLER ID = 4294967368
            runc-9466    [003] d... 292090.195348: 0: SMALLER ID = 4294967416
            runc-9466    [003] d... 292090.195419: 0: SMALLER ID = 4294967371
            runc-9466    [003] d... 292090.195489: 0: SMALLER ID = 4294968321
            runc-9466    [003] d... 292090.195590: 0: SMALLER ID = 4294967462
            runc-9466    [003] d... 292090.195663: 0: SMALLER ID = 4294967883
            runc-9466    [003] d... 292090.195749: 0: SMALLER ID = 4294969343
            runc-9466    [003] d... 292090.195826: 0: SMALLER ID = 4294967774
            runc-9466    [003] d... 292090.195942: 0: SMALLER ID = 4294967344
            runc-9466    [003] d... 292090.196026: 0: SMALLER ID = 4294968008
```

Are you sure that cgroup->kernfs_node_id->id has the same meaning in 5.4 than it has in kernels above ?

Reply:
I think the logic is broken for kernels <= 5.4:

```
# kernel 5.13 and docker in impish

$ sudo bpftool prog tracelog
systemd-1       [007] d... 630955.196985: bpf_trace_printk: BIGGER ID=31844 (docker-f19788cebc98e4d5d88a945f80e45f0ecd7652f393ec24b7c25570c1c5ed3d76.scope)

$ sudo find /sys/fs/cgroup -type d -name '*f19788cebc98e4d5d88a945f80e45f0ecd7652f393ec24b7c25570c1c5ed3d76*'
/sys/fs/cgroup/system.slice/docker-f19788cebc98e4d5d88a945f80e45f0ecd7652f393ec24b7c25570c1c5ed3d76.scope

# kernel 5.4 and docker in bionic

$ sudo cat /sys/kernel/debug/tracing/trace_pipe
runc-10231   [001] d... 292996.043605: 0: SMALLER ID=4294967516 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)
runc-10231   [001] d... 292996.045532: 0: SMALLER ID=4294967907 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)
runc-10231   [001] d... 292996.046145: 0: SMALLER ID=4294969427 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)
runc-10231   [001] d... 292996.046706: 0: SMALLER ID=4294968741 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)
runc-10231   [001] d... 292996.047514: 0: SMALLER ID=4294968048 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)
runc-10231   [001] d... 292996.047977: 0: SMALLER ID=4294968363 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)
runc-10231   [001] d... 292996.048468: 0: SMALLER ID=4294967455 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)
runc-10231   [001] d... 292996.049120: 0: SMALLER ID=4294967395 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)
runc-10231   [001] d... 292996.050122: 0: SMALLER ID=4294967359 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)
runc-10231   [001] d... 292996.050695: 0: SMALLER ID=4294967392 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)
runc-10231   [001] d... 292996.051103: 0: SMALLER ID=4294967799 (82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432)

$ sudo find /sys/fs/cgroup -type d -name '*82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432*'
/sys/fs/cgroup/hugetlb/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
/sys/fs/cgroup/cpuset/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
/sys/fs/cgroup/memory/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
/sys/fs/cgroup/pids/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
/sys/fs/cgroup/blkio/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
/sys/fs/cgroup/cpu,cpuacct/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
/sys/fs/cgroup/devices/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
/sys/fs/cgroup/net_cls,net_prio/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
/sys/fs/cgroup/perf_event/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
/sys/fs/cgroup/freezer/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
/sys/fs/cgroup/systemd/docker/82ef3a4e473f1b9b9d5be508a8a27988e78f8fde67c7bad16b5aadd27069432a
```

Reply:
I tested both on kernel 4.19 and kernel 5.10 - the logic should work.
Did you notice that I'm using the 32 LSB of the cgroup id and not the entire value?

Reply:
Nevertheless,

I was NOT able to use kconf values accepted by the verifier. While your commit 7436832 works very well with full BTF files, any attempt of having something like:

```C
extern int LINUX_KERNEL_VERSION __kconfig;
...
#ifdef CORE
    //if (bpf_core_type_exists(union kernfs_node_id)) { // <= v5.0.0
    if (LINUX_KERNEL_VERSION < KERNEL_VERSION(5, 5, 0)) {
```

causes the infamous BADRELO error code for 5.4 kernels:

```
; bpf_core_read(&kn, sizeof(void *), (struct kernfs_node *) &cgrp->kn);
360: (b7) r2 = 8
361: (bf) r3 = r7
362: (85) call bpf_probe_read#4
last_idx 362 first_idx 316
regs=4 stack=0 before 361: (bf) r3 = r7
regs=4 stack=0 before 360: (b7) r2 = 8
363: (85) call unknown#195896080
invalid func unknown#195896080 <- BAD RELO code
```

Reply:
If you keep this logic (linking cgroup ID with container_id), fixing, if needed, 5.4 scenario (if my assumption is right), then I think we should keep the type checks and work btfgen to do the relocations for us.

Reply:
> I tested both on kernel 4.19 and kernel 5.10 - the logic should work. Did you notice that I'm using the 32 LSB of the cgroup id and not the entire value?

AH okay, it makes sense, you're right. I forgot about that. So all the IDs would have the same 32 LSB cgroup ID and it would work. Okay...

Reply:
I have made btfgen to be capable of calculating the correct types in the tailored BTF file (with type based relocations feature). I'm "resolving" this conversation and starting a new one for something else.